chain/network/src/rate_limits/token_bucket.rs

```diff
@@ -0,0 +1,312 @@
+//! in order to be allowed to pass.
+//! If there aren't enough tokens in the bucket, the traffic might be stopped
+//! or delayed. However, this module responsibility stops at telling
+//! whether or not the incoming messages are allowed.
+
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use near_async::time::Duration;
+
+#[derive(thiserror::Error, Debug, PartialEq)]
+pub enum TokenBucketError {
+    #[error("invalid value for refill rate ({0})")]
+    InvalidRefillRate(f32),
+}
+
+/// Into how many parts a token can be divided.
+const TOKEN_PARTS_NUMBER: u64 = 1 << 31;
+
+/// Struct to hold the state for the token bucket algorithm. Thread-safe.
+///
+/// The precision guarantee is, at least, such that a bucket having `refill_rate` = 0.001s
+/// update at regular intervals every 10ms will successfully generate a token after 1000±1s.
+pub struct TokenBucket {
+    maximum_size: u32,
+    /// Tokens in the bucket. They are stored as `tokens * TOKEN_PARTS_NUMBER`.
+    /// In this way we can refill the bucket at shorter intervals.  
+    size: AtomicU64,
+    /// Refill rate in token per second.
+    refill_rate: f32,
+}
+
+impl TokenBucket {
+    /// Creates a new token bucket.
+    ///
+    /// # Arguments
+    ///
+    /// * `initial_size` - Initial amount of tokens in the bucket
+    /// * `maximum_size` - Maximum amount of tokens the bucket can hold
+    /// * `refill_rate` - Bucket refill rate in token per second
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if any of the arguments has an invalid value.
+    pub fn new(
+        initial_size: u32,
+        maximum_size: u32,
+        refill_rate: f32,
+    ) -> Result<Self, TokenBucketError> {
+        let size = to_tokens_with_parts(maximum_size.min(initial_size));
+        let size = AtomicU64::new(size);
+        TokenBucket::validate_refill_rate(refill_rate)?;
+        Ok(Self { maximum_size, size, refill_rate })
+    }
+
+    /// Makes an attempt to acquire `token` tokens.
+    ///
+    /// If the tokens are available they are subtracted from the current `size` and
+    /// the method returns `true`. Otherwise, `size` is not changed and the method
+    /// returns `false`.
+    pub fn acquire(&self, tokens: u32) -> bool {
+        let tokens = to_tokens_with_parts(tokens);
+        self.size
+            .fetch_update(Ordering::AcqRel, Ordering::Acquire, |size| {
+                if size >= tokens {
+                    Some(size - tokens)
+                } else {
+                    None
+                }
+            })
+            .is_ok()
+    }
+
+    /// Refills the bucket with the right number of tokens according to
+    /// the `refill_rate`, assuming an interval of time equal to `duration`
+    /// has passed. Negative durations are ignored.
+    ///
+    /// For example: if `refill_rate` == 1 and `duration` == 1s then exactly 1 token
+    /// will be added.
+    pub fn refill(&self, duration: Duration) {
```

Comment:
There is an alternative design I think is worth considering.

Imagine the TokenBucket class stores a timestamp indicating when `acquire` was last called. The `acquire` function expects an additional argument with the current time. When the internal time is advanced, the appropriate number of tokens are restored.

This allows to eliminate the `refill` function and the need to schedule regular calls to it in the PeerActor.

Reply:
Thanks for the suggestion, I thought about doing that as well. The reason I went with the `refill` function was to avoid a call to get the time when a message is received. 

Now having a look at the context where `acquire` is used I could make a small refactor and share the result of `now()` with the line 1748 `self.tracker.lock().increment_received(&self.clock, msg.len() as u64);` since both tracker and rate limits need to know the current time. 
`TokenBucket` wouldn't need to be thread safe anymore either. 

Let me know if this sounds ok with you and I'll make the change!

Reply:
Let's make the change, `clock.now()` is a very cheap call. As you mentioned we call it anyway [here](https://github.com/near/nearcore/blob/ccec393170decdc0f12cc379ce6bcab8696d4f48/chain/network/src/peer/peer_actor.rs#L1780), as well as [here](https://github.com/near/nearcore/blob/ccec393170decdc0f12cc379ce6bcab8696d4f48/chain/network/src/peer/peer_actor.rs#L1780) upon receiving a message. The benefits of eliminating the event loops to refill the buckets are much more significant here.

Reply:
I made the change, overall the functionality is the same, with the added benefit that the periodic task to refill buckets has been eliminated.